File: raytracer_rust_common/src/scene.rs
use std::f64::consts::PI;

use crate::{
    camera::{Background, CameraSceneOptions},
    components::{
        material::{Dielectric, DiffuseLight, Isotropic, Lambert, Material, MaterialEnum, Metal},
        noise::Perlin,
        surface::{
            Surface,
            primitive::{Quad, Sphere},
            structure::{BoundingVolumeHierarchy, SurfaceList},
            volume::ConstantMedium,
        },
        texture::{Checker, NoiseTexture, Texture},
    },
    util::{
        interval::Interval,
        vec3::{Color, arrow, color, point},
    },
};

#[derive(Debug, Clone)]
pub struct Scene {
    pub world: BoundingVolumeHierarchy,
    pub camera: CameraSceneOptions,
}

#[derive(Debug, Clone, Copy, Default)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub enum BuiltinScene {
    Empty,
    #[default]
    First,
    Second,
    MovingSpheres,
    CheckeredSpheres,
    PerlinSpheres,
    Quads,
    SimpleLight,
    CornellBox,
    CornellSmoke,
}

impl BuiltinScene {
    pub fn to_scene(self) -> Scene {
        let mut world = SurfaceList::new();
        let camera = match self {
            BuiltinScene::Empty => CameraSceneOptions::default(),

            BuiltinScene::First => {
                let material_ground = Lambert::new(color(0.8, 0.8, 0.0));
                let material_center = Lambert::new(color(0.1, 0.2, 0.5));
                let material_left = Dielectric::new(1.50);
                let material_bubble = Dielectric::new(1.00 / 1.50);
                let material_right = Metal::new(color(0.8, 0.6, 0.2), 1.0);

                world.add(Sphere::stationary(
                    point(0.0, -100.5, -10.0),
                    100.0,
                    material_ground,
                ));
                world.add(Sphere::stationary(
                    point(0.0, 0.0, -1.2),
                    0.5,
                    material_center,
                ));

                world.add(Sphere::stationary(
                    point(-1.0, 0.0, -1.0),
                    0.5,
                    material_left,
                ));
                world.add(Sphere::stationary(
                    point(-1.0, 0.0, -1.0),
                    0.4,
                    material_bubble,
                ));
                world.add(Sphere::stationary(
                    point(1.0, 0.0, -1.0),
                    0.5,
                    material_right,
                ));

                CameraSceneOptions {
                    vertical_fov: 20.0,
                    look_from: point(-2.0, 2.0, 1.0),
                    ..Default::default()
                }
            }

            BuiltinScene::Second => {
                let r = (PI / 4.0).cos();

                let material_left = Lambert::new(color(0.0, 0.0, 1.0));
                let material_right = Lambert::new(color(1.0, 0.0, 0.0));

                world.add(Sphere::stationary(point(-r, 0.0, -1.0), r, material_left));
                world.add(Sphere::stationary(point(r, 0.0, -1.0), r, material_right));

                CameraSceneOptions {
                    vertical_fov: 90.0,
                    ..Default::default()
                }
            }

            BuiltinScene::MovingSpheres => {
                let material_ground = Lambert::new(Checker::new(
                    0.32,
                    color(0.2, 0.3, 0.1),
                    color(0.9, 0.9, 0.9),
                ));
                world.add(Sphere::stationary(
                    point(0.0, -1000.0, 0.0),
                    1000.0,
                    material_ground,
                ));

                for a in -11..11 {
                    for b in -11..11 {
                        let choose_material = Interval::UNIT.random_double();
                        let center_start = point(
                            a as f64 + 0.9 * Interval::UNIT.random_double(),
                            0.2,
                            b as f64 + 0.9 * Interval::UNIT.random_double(),
                        );
                        if (center_start - point(4.0, 0.2, 0.0)).length() <= 0.9 {
                            continue;
                        }

                        let material: MaterialEnum = if choose_material < 0.8 {
                            let albedo =
                                Color::random(Interval::UNIT) * Color::random(Interval::UNIT);
                            Lambert::new(albedo).into()
                        } else if choose_material < 0.95 {
                            let albedo =
                                Color::random(Interval::UNIT) * Color::random(Interval::UNIT);
                            let fuzz = Interval::HALF.random_double();
                            Metal::new(albedo, fuzz).into()
                        } else {
                            Dielectric::new(1.5).into()
                        };

                        // if Interval::UNIT.random_double() < 0.5 {
                        if false {
                            let center_end =
                                center_start + point(0.0, Interval::HALF.random_double(), 0.0);
                            world.add(Sphere::moving(center_start, center_end, 0.2, material));
                        } else {
                            world.add(Sphere::stationary(center_start, 0.2, material));
                        }
                    }
                }

                let material1 = Dielectric::new(1.5);
                world.add(Sphere::stationary(point(0.0, 1.0, 0.0), 1.0, material1));

                let material2 = Lambert::new(color(0.4, 0.2, 0.1));
                world.add(Sphere::stationary(point(-4.0, 1.0, 0.0), 1.0, material2));

                let material3 = Metal::new(color(0.7, 0.6, 0.5), 0.0);
                world.add(Sphere::stationary(point(4.0, 1.0, 0.0), 1.0, material3));

                CameraSceneOptions {
                    vertical_fov: 20.0,
                    look_from: point(13.0, 2.0, 3.0),
                    look_at: point(0.0, 0.0, 0.0),
                    defocus_angle: 0.6,
                    focus_distance: 10.0,
                    ..Default::default()
                }
            }

            BuiltinScene::CheckeredSpheres => {
                let material = Lambert::new(Checker::new(
                    0.32,
                    color(0.2, 0.3, 0.1),
                    color(0.9, 0.9, 0.9),
                ))
                .shared();

                world.add(Sphere::stationary(
                    point(0.0, -10.0, 0.0),
                    10.0,
                    material.clone(),
                ));
                world.add(Sphere::stationary(
                    point(0.0, 10.0, 0.0),
                    10.0,
                    material.clone(),
                ));

                CameraSceneOptions {
                    vertical_fov: 20.0,
                    look_from: point(13.0, 2.0, 3.0),
                    look_at: point(0.0, 0.0, 0.0),
                    ..Default::default()
                }
            }

            BuiltinScene::PerlinSpheres => {
                let material = Lambert::new(NoiseTexture::new(Perlin::new(), 4.0)).shared();
                world.add(Sphere::stationary(
                    point(0.0, -1000.0, 0.0),
                    1000.0,
                    material.clone(),
                ));
                world.add(Sphere::stationary(point(0.0, 2.0, 0.0), 2.0, material));

                CameraSceneOptions {
                    vertical_fov: 20.0,
                    look_from: point(13.0, 2.0, 3.0),
                    look_at: point(0.0, 0.0, 0.0),
                    ..Default::default()
                }
            }

            BuiltinScene::Quads => {
                world.add(Quad::new(
                    point(-3.0, -2.0, 5.0),
                    arrow(0.0, 0.0, -4.0),
                    arrow(0.0, 4.0, 0.0),
                    Lambert::new(color(1.0, 0.2, 0.2)),
                ));
                world.add(Quad::new(
                    point(-2.0, -2.0, 0.0),
                    arrow(4.0, 0.0, 0.0),
                    arrow(0.0, 4.0, 0.0),
                    Lambert::new(color(0.2, 1.0, 0.2)),
                ));
                world.add(Quad::new(
                    point(3.0, -2.0, 1.0),
                    arrow(0.0, 0.0, 4.0),
                    arrow(0.0, 4.0, 0.0),
                    Lambert::new(color(0.2, 0.2, 1.0)),
                ));
                world.add(Quad::new(
                    point(-2.0, 3.0, 1.0),
                    arrow(4.0, 0.0, 0.0),
                    arrow(0.0, 0.0, 4.0),
                    Lambert::new(color(1.0, 0.5, 0.0)),
                ));
                world.add(Quad::new(
                    point(-2.0, -3.0, 5.0),
                    arrow(4.0, 0.0, 0.0),
                    arrow(0.0, 0.0, -4.0),
                    Lambert::new(color(0.2, 0.8, 0.8)),
                ));

                CameraSceneOptions {
                    vertical_fov: 80.0,
                    look_from: point(0.0, 0.0, 9.0),
                    look_at: point(0.0, 0.0, 0.0),
                    ..Default::default()
                }
            }

            BuiltinScene::SimpleLight => {
                let material_perlin = Lambert::new(NoiseTexture::new(Perlin::new(), 4.0).shared());
                world.add(Sphere::stationary(
                    point(0.0, -1000.0, 0.0),
                    1000.0,
                    material_perlin.clone(),
                ));
                world.add(Sphere::stationary(
                    point(0.0, 2.0, 0.0),
                    2.0,
                    material_perlin,
                ));

                let material_light = DiffuseLight::new(color(4.0, 4.0, 4.0)).shared();
                world.add(Sphere::stationary(
                    point(0.0, 7.0, 0.0),
                    2.0,
                    material_light.clone(),
                ));
                world.add(Quad::new(
                    point(3.0, 1.0, -2.0),
                    arrow(2.0, 0.0, 0.0),
                    arrow(0.0, 2.0, 0.0),
                    material_light,
                ));

                CameraSceneOptions {
                    vertical_fov: 20.0,
                    look_from: point(26.0, 3.0, 6.0),
                    look_at: point(0.0, 2.0, 0.0),
                    background: Background::Solid(color(0.0, 0.0, 0.0)),
                    ..Default::default()
                }
            }

            BuiltinScene::CornellBox => {
                let material_red = Lambert::new(color(0.65, 0.05, 0.05));
                let material_white = Lambert::new(color(0.73, 0.73, 0.73)).shared();
                let material_green = Lambert::new(color(0.12, 0.45, 0.15));
                let material_light = DiffuseLight::new(color(15.0, 15.0, 15.0));

                world.add(Quad::new(
                    point(555.0, 0.0, 0.0),
                    arrow(0.0, 555.0, 0.0),
                    arrow(0.0, 0.0, 555.0),
                    material_green,
                ));
                world.add(Quad::new(
                    point(0.0, 0.0, 0.0),
                    arrow(0.0, 555.0, 0.0),
                    arrow(0.0, 0.0, 555.0),
                    material_red,
                ));
                world.add(Quad::new(
                    point(343.0, 554.0, 332.0),
                    arrow(-130.0, 0.0, 0.0),
                    arrow(0.0, 0.0, -105.0),
                    material_light,
                ));
                world.add(Quad::new(
                    point(0.0, 0.0, 0.0),
                    arrow(555.0, 0.0, 0.0),
                    arrow(0.0, 0.0, 555.0),
                    material_white.clone(),
                ));
                world.add(Quad::new(
                    point(555.0, 555.0, 555.0),
                    arrow(-555.0, 0.0, 0.0),
                    arrow(0.0, 0.0, -555.0),
                    material_white.clone(),
                ));
                world.add(Quad::new(
                    point(0.0, 0.0, 555.0),
                    arrow(555.0, 0.0, 0.0),
                    arrow(0.0, 555.0, 0.0),
                    material_white.clone(),
                ));

                world.add(
                    Quad::cube(
                        point(0.0, 0.0, 0.0),
                        point(165.0, 330.0, 165.0),
                        material_white.clone(),
                    )
                    .rotate_y(15.0)
                    .translate(arrow(265.0, 0.0, 295.0)),
                );
                world.add(
                    Quad::cube(
                        point(0.0, 0.0, 0.0),
                        point(165.0, 165.0, 165.0),
                        material_white,
                    )
                    .rotate_y(-18.0)
                    .translate(arrow(130.0, 0.0, 65.0)),
                );

                CameraSceneOptions {
                    vertical_fov: 40.0,
                    look_from: point(278.0, 278.0, -800.0),
                    look_at: point(278.0, 278.0, 0.0),
                    background: Background::Solid(color(0.0, 0.0, 0.0)),
                    ..Default::default()
                }
            }

            BuiltinScene::CornellSmoke => {
                let material_red = Lambert::new(color(0.65, 0.05, 0.05));
                let material_white = Lambert::new(color(0.73, 0.73, 0.73)).shared();
                let material_green = Lambert::new(color(0.12, 0.45, 0.15));
                let material_light = DiffuseLight::new(color(7.0, 7.0, 7.0));

                world.add(Quad::new(
                    point(555.0, 0.0, 0.0),
                    arrow(0.0, 555.0, 0.0),
                    arrow(0.0, 0.0, 555.0),
                    material_green,
                ));
                world.add(Quad::new(
                    point(0.0, 0.0, 0.0),
                    arrow(0.0, 555.0, 0.0),
                    arrow(0.0, 0.0, 555.0),
                    material_red,
                ));
                world.add(Quad::new(
                    point(113.0, 554.0, 127.0),
                    arrow(330.0, 0.0, 0.0),
                    arrow(0.0, 0.0, 305.0),
                    material_light,
                ));
                world.add(Quad::new(
                    point(0.0, 0.0, 0.0),
                    arrow(555.0, 0.0, 0.0),
                    arrow(0.0, 0.0, 555.0),
                    material_white.clone(),
                ));
                world.add(Quad::new(
                    point(555.0, 555.0, 555.0),
                    arrow(-555.0, 0.0, 0.0),
                    arrow(0.0, 0.0, -555.0),
                    material_white.clone(),
                ));
                world.add(Quad::new(
                    point(0.0, 0.0, 555.0),
                    arrow(555.0, 0.0, 0.0),
                    arrow(0.0, 555.0, 0.0),
                    material_white.clone(),
                ));

                world.add(ConstantMedium::new(
                    Quad::cube(
                        point(0.0, 0.0, 0.0),
                        point(165.0, 330.0, 165.0),
                        material_white.clone(),
                    )
                    .rotate_y(15.0)
                    .translate(arrow(265.0, 0.0, 295.0)),
                    0.01,
                    Isotropic::new(color(0.0, 0.0, 0.0)),
                ));
                world.add(ConstantMedium::new(
                    Quad::cube(
                        point(0.0, 0.0, 0.0),
                        point(165.0, 165.0, 165.0),
                        material_white,
                    )
                    .rotate_y(-18.0)
                    .translate(arrow(130.0, 0.0, 65.0)),
                    0.01,
                    Isotropic::new(color(1.0, 1.0, 1.0)),
                ));

                CameraSceneOptions {
                    vertical_fov: 40.0,
                    look_from: point(278.0, 278.0, -800.0),
                    look_at: point(278.0, 278.0, 0.0),
                    background: Background::Solid(color(0.0, 0.0, 0.0)),
                    ..Default::default()
                }
            }
        };

        Scene {
            world: world.into(),
            camera,
        }
    }
}

File: raytracer_ts/scene.ts
import { type CameraSceneOptions } from './camera';
import { BoundingVolumeHierarchy, Hittable, HittableList, Quad, Sphere } from './hittable';
import { Dielectric, DiffuseLight, Lambert, Metal } from './material';
import { Perlin } from './noise';
import { Checker, NoiseTexture, SolidColor, TurbulenceTexture } from './texture';
import { Interval, Vec3 } from './util';

class Scene {
    constructor(
        public world: Hittable,
        public cameraOptions: CameraSceneOptions,
    ) {}
}

export const scene1 = () => {
    const materialGround = new Lambert(new SolidColor(new Vec3(0.8, 0.8, 0.0)));
    const materialCenter = new Lambert(new SolidColor(new Vec3(0.1, 0.2, 0.5)));
    const materialLeft = new Dielectric(1.5);
    const materialBubble = new Dielectric(1.0 / 1.5);
    const materialRight = new Metal(new Vec3(0.8, 0.6, 0.2), 1.0);

    return new Scene(
        new HittableList([
            Sphere.stationary(new Vec3(0.0, -100.5, -1.0), 100.0, materialGround),
            Sphere.stationary(new Vec3(0.0, 0.0, -1.2), 0.5, materialCenter),
            Sphere.stationary(new Vec3(-1.0, 0.0, -1.0), 0.5, materialLeft),
            Sphere.stationary(new Vec3(-1.0, 0.0, -1.0), 0.4, materialBubble),
            Sphere.stationary(new Vec3(1.0, 0.0, -1.0), 0.5, materialRight),
        ]),
        {
            verticalFov: 20,
            lookFrom: new Vec3(-2, 2, 1),
            lookAt: new Vec3(0, 0, -1),
            vUp: new Vec3(0, 1, 0),

            defocusAngle: 10.0,
            focusDistance: 3.4,
        },
    );
};

export const scene2 = () => {
    const R = Math.cos(Math.PI / 4);

    const materialLeft = new Lambert(new SolidColor(new Vec3(0, 0, 1)));
    const materialRight = new Lambert(new SolidColor(new Vec3(1, 0, 0)));

    return new Scene(
        new HittableList([
            Sphere.stationary(new Vec3(-R, 0, -1), R, materialLeft),
            Sphere.stationary(new Vec3(R, 0, -1), R, materialRight),
        ]),
        {
            verticalFov: 90,
            lookFrom: new Vec3(0, 0, 0),
            lookAt: new Vec3(0, 0, -1),
            vUp: new Vec3(0, 1, 0),

            defocusAngle: 0,
            focusDistance: 10,
        },
    );
};

const randomMaterial = () => {
    const chooseMaterial = Interval.unit.random();

    if (chooseMaterial < 0.8) {
        const albedo = Vec3.random(Interval.unit).times(Vec3.random(Interval.unit));
        return new Lambert(new SolidColor(albedo));
    } else if (chooseMaterial < 0.95) {
        const albedo = Vec3.random(new Interval(0, 0.5));
        const fuzz = new Interval(0, 0.5).random();
        return new Metal(albedo, fuzz);
    } else {
        return new Dielectric(1.5);
    }
};

export const movingSpheres = () => {
    const world = new HittableList();

    const groundMaterial = new Lambert(
        new Checker(
            0.32, //
            new SolidColor(new Vec3(0.2, 0.3, 0.1)),
            new SolidColor(new Vec3(0.9, 0.9, 0.9)),
        ),
    );
    world.add(Sphere.stationary(new Vec3(0, -1000, 0), 1000, groundMaterial));

    for (let a = -11; a < 11; a++) {
        for (let b = -11; b < 11; b++) {
            const center = new Vec3(a + 0.9 * Interval.unit.random(), 0.2, b + 0.9 * Interval.unit.random());

            if (center.minus(new Vec3(4, 0.2, 0)).length > 0.9) {
                const endCenter = center.plus(new Vec3(0, new Interval(0, 0.5).random(), 0));
                world.add(Sphere.moving(center, endCenter, 0.2, randomMaterial()));
            }
        }
    }

    const material1 = new Dielectric(1.5);
    world.add(Sphere.stationary(new Vec3(0, 1, 0), 1.0, material1));

    const material2 = new Lambert(new SolidColor(new Vec3(0.4, 0.2, 0.1)));
    world.add(Sphere.stationary(new Vec3(-4, 1, 0), 1.0, material2));

    const material3 = new Metal(new Vec3(0.7, 0.6, 0.5), 0.0);
    world.add(Sphere.stationary(new Vec3(4, 1, 0), 1.0, material3));

    return new Scene(world.toBVH(), {
        verticalFov: 20,
        lookFrom: new Vec3(13, 2, 3),
        lookAt: new Vec3(0, 0, 0),
        vUp: new Vec3(0, 1, 0),

        defocusAngle: 0.6,
        focusDistance: 10,
    });
};

export const checkeredSpheres = () => {
    const world = new HittableList();

    const checker = new Checker(0.32, new SolidColor(new Vec3(0.2, 0.3, 0.1)), new SolidColor(new Vec3(0.9, 0.9, 0.9)));
    world.add(Sphere.stationary(new Vec3(0, -10, 0), 10, new Lambert(checker)));
    world.add(Sphere.stationary(new Vec3(0, 10, 0), 10, new Lambert(checker)));

    return new Scene(world.toBVH(), {
        verticalFov: 20,
        lookFrom: new Vec3(13, 2, 3),
        lookAt: new Vec3(0, 0, 0),
        vUp: new Vec3(0, 1, 0),

        defocusAngle: 0,
        focusDistance: 10,
    });
};

export const perlinSpheres = () => {
    const world = new HittableList();

    const perlin = new TurbulenceTexture(new Perlin(), 4);
    world.add(Sphere.stationary(new Vec3(0, -1000, 0), 1000, new Lambert(perlin)));
    world.add(Sphere.stationary(new Vec3(0, 2, 0), 2, new Lambert(perlin)));

    return new Scene(world.toBVH(), {
        verticalFov: 20,
        lookFrom: new Vec3(13, 2, 3),
        lookAt: new Vec3(0, 0, 0),
        vUp: new Vec3(0, 1, 0),

        defocusAngle: 0,
        focusDistance: 10,
    });
};

export const quads = () => {
    const data = [
        [new Vec3(-3, -2, 5), new Vec3(0, 0, -4), new Vec3(0, 4, 0), new Vec3(1.0, 0.2, 0.2)],
        [new Vec3(-2, -2, 0), new Vec3(4, 0, 0), new Vec3(0, 4, 0), new Vec3(0.2, 1.0, 0.2)],
        [new Vec3(3, -2, 1), new Vec3(0, 0, 4), new Vec3(0, 4, 0), new Vec3(0.2, 0.2, 1.0)],
        [new Vec3(-2, 3, 1), new Vec3(4, 0, 0), new Vec3(0, 0, 4), new Vec3(1.0, 0.5, 0.0)],
        [new Vec3(-2, -3, 5), new Vec3(4, 0, 0), new Vec3(0, 0, -4), new Vec3(0.2, 0.8, 0.8)],
    ];

    const world = new BoundingVolumeHierarchy(
        data.map(([Q, u, v, albedo]) => new Quad(Q, u, v, new Lambert(new SolidColor(albedo)))),
    );

    return new Scene(world, {
        verticalFov: 80,
        lookFrom: new Vec3(0, 0, 9),
        lookAt: new Vec3(0, 0, 0),
        vUp: new Vec3(0, 1, 0),
        defocusAngle: 0,
        focusDistance: 10,
    });
};

export const simpleLight = () => {
    const world = new HittableList();

    const perlin = new NoiseTexture(new Perlin(), 4);
    world.add(Sphere.stationary(new Vec3(0, -1000, 0), 1000, new Lambert(perlin)));
    world.add(Sphere.stationary(new Vec3(0, 2, 0), 2, new Lambert(perlin)));

    const light = new DiffuseLight(new SolidColor(new Vec3(10, 10, 10)));
    world.add(Sphere.stationary(new Vec3(0, 7, 0), 2, light));
    world.add(new Quad(new Vec3(3, 1, -2), new Vec3(2, 0, 0), new Vec3(0, 2, 0), light));

    return new Scene(world.toBVH(), {
        verticalFov: 20,
        lookFrom: new Vec3(26, 3, 6),
        lookAt: new Vec3(0, 2, 0),
        vUp: new Vec3(0, 1, 0),
        defocusAngle: 0,
        focusDistance: 10,
        background: new Vec3(0, 0, 0),
    });
};

export const cornellBox = () => {
    const red = new Lambert(new SolidColor(new Vec3(0.65, 0.05, 0.05)));
    const white = new Lambert(new SolidColor(new Vec3(0.73, 0.73, 0.73)));
    const green = new Lambert(new SolidColor(new Vec3(0.12, 0.45, 0.15)));
    const light = new DiffuseLight(new SolidColor(new Vec3(15, 15, 15)));

    const quads = [
        [new Vec3(555, 0, 0), new Vec3(0, 555, 0), new Vec3(0, 0, 555), green],
        [new Vec3(0, 0, 0), new Vec3(0, 555, 0), new Vec3(0, 0, 555), red],
        [new Vec3(343, 554, 332), new Vec3(-130, 0, 0), new Vec3(0, 0, -105), light],
        [new Vec3(0, 0, 0), new Vec3(555, 0, 0), new Vec3(0, 0, 555), white],
        [new Vec3(555, 555, 555), new Vec3(-555, 0, 0), new Vec3(0, 0, -555), white],
        [new Vec3(0, 0, 555), new Vec3(555, 0, 0), new Vec3(0, 555, 0), white],
    ] as const;

    const world = new HittableList(quads.map(([Q, u, v, material]) => new Quad(Q, u, v, material)));

    world.add(Quad.box(new Vec3(130, 0, 65), new Vec3(295, 165, 230), white));
    world.add(Quad.box(new Vec3(265, 0, 295), new Vec3(430, 330, 460), white));

    return new Scene(world.toBVH(), {
        verticalFov: 40,
        lookFrom: new Vec3(278, 278, -800),
        lookAt: new Vec3(278, 278, 0),
        vUp: new Vec3(0, 1, 0),
        defocusAngle: 0,
        focusDistance: 10,
        background: new Vec3(0, 0, 0),
    });
};
